### PR TITLE
criterion: init at 2.3.3

### PR DESCRIPTION
**Pull request by Yumasi**
###### Motivation for this change

`Criterion` was not packaged for NixOS, so here is a derivation for it. Alongside it is a derivation for `BoxFort`, which is a library specifically created to implement `Criterion`. Since `BoxFort` is a very nice project, it does not have any release and is built by Criterion's build system against `BoxFort`'s master branch. Since this behavior is not Nix-friendly, I packaged it using the commit hash as a version number. This also guarantees me to have a working Criterion derivation. This PR also adds me in the maintainer list for those packages.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/boxfort/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  src = fetchFromGitHub {
+    owner = "Snaipe";
+    repo = "BoxFort";
+    rev = "926bd4ce968592dbbba97ec1bb9aeca3edf29b0d";
+    sha256 = "0mzy4f8qij6ckn5578y3l4rni2470pdkjy5xww7ak99l1kh3p3v6";
+  };
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [ cmake pkgconfig gettext libcsptr dyncall nanomsg ];
```
Comment (wucke13) [inline pkgs/development/libraries/boxfort/default.nix]: Are all of these `nativeBuildInputs`? For example I have doubts that `libcsptr` is a `nativeBuildInput`. Please refer to the [manual](https://nixos.org/nixpkgs/manual/#ssec-stdenv-dependencies) for further information.

File: pkgs/development/libraries/boxfort/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  cmakeFlags = [ "-DBXF_TESTS=OFF" "-DBXF_SAMPLES=OFF" "-DBXF_FORK_RESILIENCE=OFF" ];
+  doCheck = false;
```
Comment (wucke13) [inline pkgs/development/libraries/boxfort/default.nix]: Why is it that the checks fail?

File: pkgs/development/libraries/boxfort/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
+{ stdenv, fetchFromGitHub, cmake, pkgconfig, gettext, libcsptr, dyncall, nanomsg}:
+
```
Comment (wucke13) [inline pkgs/development/libraries/boxfort/default.nix]: I would prefer `pkg-config` over `pkgconfig`, as the latter is only an alias of the former, though that might be considered cosmetics...

File: pkgs/development/libraries/criterion/default.nix [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  ];
+
+  checkInputs = with python37Packages; [
+    cram
+  ];
+  cmakeFlags = [ "-DCTESTS=ON" ];
+  doCheck = true;
+  preCheck = ''
+    export LD_LIBRARY_PATH=`pwd`:$LD_LIBRARY_PATH
+    '';
```
Comment (wucke13) [inline pkgs/development/libraries/criterion/default.nix]: Fix spacing: remove two spaces please

**Comment by wucke13** [comment]
I guess you forgot to add your package to `pkgs/top-level/all-packages.nix`? On reviewing your PR
(`nix-review pr 70403`) the result is that no packages have been build. Please look into this :)

**Comment by Yumasi** [comment]
@wucke13 Thanks for your review! I've pushed updated commits with your comments taken into account.

**Comment by Thesola10** [comment]
Hello, I am the author of #70609 which also adds `criterion` to Nix packages. Since our PRs are conflicting, I wanted to ask if you are okay with my PR, or if I should close it and have yours be merged.

**Comment by Yumasi** [comment]
Hi ! I guess its a matter of who gets to maintain it. You seem to have two extra commits in your pull-request and, purely subjectively, I find my derivation to be cleaner, so I'd say my PR is currently more ready for merge.

File: pkgs/development/libraries/boxfort/default.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
+{ stdenv, fetchFromGitHub, cmake, pkg-config, gettext, libcsptr, dyncall
+, nanomsg, python37Packages }:
+
+stdenv.mkDerivation rec {
+  name = "boxfort";
```
Comment (Thesola10) [inline pkgs/development/libraries/boxfort/default.nix]: You should use `pname` and specify `unstable-2019-09-19` as a version so as to follow Nix naming conventions.

**Comment by Thesola10** [comment]
I think I actually agree with you on this, I was just working on rebasing these two (accidental) commits away, but your derivation file is more complete and cleaner. (though I can't help but wonder if running unit tests in a package build is really necessary)

I will be closing my PR shortly.

**Comment by wucke13** [comment]
@Thesola10 @Yumasi 

> I guess its a matter of who gets to maintain it

Noo! If you want to, add both of you to the maintainer list. It is never wrong to have more people to fix things once they go bad!

**Comment by Yumasi** [comment]
> Noo! If you want to, add both of you to the maintainer list. It is never wrong to have more people to fix things once they go bad!

I couldn't agree more! @Thesola10 do you want me to add you to the maintainer list ?

**Comment by Thesola10** [comment]
@Yumasi please do it then (my maintainer entry already exists and is `thesola10`) because I don't want to mess up this PR branch. I don't even think I have access to this patch repo.

File: maintainers/maintainer-list.nix [hunk 6]
```diff
@@ -7152,6 +7152,12 @@
     githubId = 1866448;
     name = "Eric Bailey";
   };
+  Yumasi = {
+    email = "gpagnoux@gmail.com";
+    github = "Yumasi";
+    githubId = 24368641;
+    name = "Guillaume Pagnoux";
+  };
```
Comment (Thesola10) [inline maintainers/maintainer-list.nix]: You can add a public key (which I see you already have to sign your commits) in the format:

```nix
keys = [{
  longkeyid = "keytype/0xFINGERPRINT";
  fingerprint = "FINGERPRINT with spaces each 4 hexadigits";
}];

Comment (Thesola10) [reply maintainers/maintainer-list.nix]: Also, consider uploading your public key to a PGP keyserver

Comment (Yumasi) [reply maintainers/maintainer-list.nix]: Thanks for the suggestion, I added it. Actually, it is on a PGP keyserver already (pgp.mit.edu)

**Comment by Thesola10** [comment]
that's cool, but why exactly did ofborg request me for review?

**Review by Thesola10 — APPROVED** [review]
Looks good to me.

**Comment by Yumasi** [comment]
> that's cool, but why exactly did ofborg request me for review?

Probably because I added you to the package maintainer list. Thanks for your reviews!

Anyway, this look good to me as well, so I think we can get this merged at this point, if everyone is ok with it. @wucke13 is there anything else you think I should change ?

**Review by wucke13 — APPROVED** [review]
Looks good to me.
Not necessary but something about the cosmetics again:
I prefer to have my deps sorted by nativeBuildInputs (`stdenv`, `cmake`, ...), then buildInputs (`libxyz`, ...) and then check and postBuild deps. All of these groups sorted by alphanumerical order (but stripping away very generic prefixes like `lib`. That is just a personal preference, but I think it looks nice plus deps are named in the same order in the arguments and the vars, so that checking for deps which are mentioned in the arguments but not mentioned anywhere in the build are easier to spot by eye.

**Comment by aanderse** [comment]
@GrahamcOfBorg build boxfort criterion

**Comment by Thesola10** [comment]
can't wait for it to reach `nixpkgs`, I'm using my "homemade" Criterion derivation in a Nix shell, having a more official package will simplify things 🎉

**Pull request merged**